src/backend/access/transam/xlog.c

```diff
@@ -8142,6 +8144,8 @@ StartupXLOG(void)
 	 */
 	if (fast_promoted)
 		RequestCheckpoint(CHECKPOINT_FORCE);
+
+	*shmCleanupBackends = true;
```

Comment:
Why do we need setting this flag conditionally above and unconditionally here?

Reply:
The first time - during a crash restart, and the second time - during a normal restart (restart only the coordinator, for example, manually via pg_ctl restart)

Reply:
Will we not reach the 2nd case in case of crash recovery?

Reply:
Yes, if some mistake happens before, for example, https://github.com/arenadata/gpdb/pull/567/files#diff-c1cb3ab2a19606390c1a7ed00ffe5a45531702ca5faf999d401c548f8951c65bR6733-R6736

Reply:
But only with the second change both tests pass.

Reply:
Dtx Recovery starts only on the coordinator, which is defined in the DtxRecoveryStartRule function.

Reply:
Will  `*shmCleanupBackends = true` be reached for all startup scenarios (including hot standby launching as Stolb27 has asked)?

Reply:
Because since I removed the first assignment, it won't work for hot standby.